File: backend/app/models/expense.py
import uuid
from datetime import date as date_type
from datetime import datetime
from decimal import Decimal

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base
from app.models.base import TimestampMixin, UUIDMixin


class Expense(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "expenses"

    group_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("groups.id", ondelete="CASCADE"), nullable=False
    )
    # Optional link to the originating Plaid/manual transaction
    transaction_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("transactions.id", ondelete="SET NULL"), nullable=True
    )
    # Dedup key for Splitwise-synced rows; null for self-hosted-only expenses
    splitwise_expense_id: Mapped[str | None] = mapped_column(
        String(64), unique=True, nullable=True
    )
    description: Mapped[str] = mapped_column(String(512), nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="USD")
    date: Mapped[date_type] = mapped_column(Date, nullable=False)
    category: Mapped[str | None] = mapped_column(String(128), nullable=True)
    # Provenance from Splitwise (user_identifiers + the real added/edited timestamps, distinct from our
    # row's created_at/updated_at which track import time); null for self-hosted.
    created_by: Mapped[str | None] = mapped_column(String(128), nullable=True)
    updated_by: Mapped[str | None] = mapped_column(String(128), nullable=True)
    splitwise_created_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    splitwise_updated_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    comments_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    # Recurrence/bundle (display labels only — each occurrence still arrives as its own expense).
    repeats: Mapped[bool | None] = mapped_column(Boolean, nullable=True)
    repeat_interval: Mapped[str | None] = mapped_column(String(32), nullable=True)
    expense_bundle_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    # Splitwise receipt image URL (remote, not our proxied bytes) + simplified repayments, both from import.
    splitwise_receipt_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    repayments: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    # Soft-delete marker; null = active (DELETE archives unless hard-delete is enabled)
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    group: Mapped["Group"] = relationship(back_populates="expenses")  # noqa: F821
    items: Mapped[list["ExpenseItem"]] = relationship(  # noqa: F821
        back_populates="expense", cascade="all, delete-orphan"
    )
    splits: Mapped[list["Split"]] = relationship(  # noqa: F821
        back_populates="expense", cascade="all, delete-orphan"
    )
    receipts: Mapped[list["Receipt"]] = relationship(  # noqa: F821
        back_populates="expense", cascade="all, delete-orphan"
    )

File: ios/SplitBackAPI/Sources/SplitBackAPI/Models/Expense.swift
import Foundation
import SwiftData

/// An expense in a group, with its splits, line items, and receipts.
/// Mirrors the server `expenses` table. Dedupe key for Splitwise-sourced rows is
/// `splitwiseExpenseId`.
@Model
final class Expense {
    @Attribute(.unique) var id: UUID
    var groupId: UUID
    var transactionId: UUID?
    var splitwiseExpenseId: String?
    var details: String
    var amount: Decimal
    var currency: String
    var date: Date
    var category: String?
    /// Splitwise provenance: who added/edited it and the real added/edited timestamps (distinct from
    /// createdAt/updatedAt, which track when we imported the row). Plus notes, comments, recurrence.
    var createdByIdentifier: String?
    var updatedByIdentifier: String?
    var splitwiseCreatedAt: Date?
    var splitwiseUpdatedAt: Date?
    var notes: String?
    var commentsCount: Int?
    var repeats: Bool?
    var repeatInterval: String?
    var expenseBundleId: String?
    /// Splitwise receipt image URL (remote) + the simplified repayments as a raw JSON string, from import.
    var splitwiseReceiptURL: String?
    var splitwiseRepayments: String?
    /// The caller's per-user budget overrides (from `expense_overrides`); null = the default (included).
    /// Excludes this expense's owed-share from spending / cash-flow analytics without changing balances.
    var includeInSpending: Bool?
    var includeInCashFlow: Bool?
    var createdAt: Date
    var updatedAt: Date

    @Relationship(deleteRule: .cascade, inverse: \Split.expense)
    var splits: [Split]
    @Relationship(deleteRule: .cascade, inverse: \ExpenseItem.expense)
    var items: [ExpenseItem]
    @Relationship(deleteRule: .cascade, inverse: \Receipt.expense)
    var receipts: [Receipt]

    init(
        id: UUID,
        groupId: UUID,
        transactionId: UUID? = nil,
        splitwiseExpenseId: String? = nil,
        details: String,
        amount: Decimal,
        currency: String,
        date: Date,
        category: String? = nil,
        createdByIdentifier: String? = nil,
        updatedByIdentifier: String? = nil,
        splitwiseCreatedAt: Date? = nil,
        splitwiseUpdatedAt: Date? = nil,
        notes: String? = nil,
        commentsCount: Int? = nil,
        repeats: Bool? = nil,
        repeatInterval: String? = nil,
        expenseBundleId: String? = nil,
        splitwiseReceiptURL: String? = nil,
        splitwiseRepayments: String? = nil,
        includeInSpending: Bool? = nil,
        includeInCashFlow: Bool? = nil,
        createdAt: Date,
        updatedAt: Date,
        splits: [Split] = [],
        items: [ExpenseItem] = [],
        receipts: [Receipt] = []
    ) {
        self.id = id
        self.groupId = groupId
        self.transactionId = transactionId
        self.splitwiseExpenseId = splitwiseExpenseId
        self.details = details
        self.amount = amount
        self.currency = currency
        self.date = date
        self.category = category
        self.createdByIdentifier = createdByIdentifier
        self.updatedByIdentifier = updatedByIdentifier
        self.splitwiseCreatedAt = splitwiseCreatedAt
        self.splitwiseUpdatedAt = splitwiseUpdatedAt
        self.notes = notes
        self.commentsCount = commentsCount
        self.repeats = repeats
        self.repeatInterval = repeatInterval
        self.expenseBundleId = expenseBundleId
        self.splitwiseReceiptURL = splitwiseReceiptURL
        self.splitwiseRepayments = splitwiseRepayments
        self.includeInSpending = includeInSpending
        self.includeInCashFlow = includeInCashFlow
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.splits = splits
        self.items = items
        self.receipts = receipts
    }
}
